Change rememberRetain not to retain the value of removed node

vulpeszerda opened the pull request:
As mentioned in #1783, I propose modifying the behavior of `rememberRetained` to improve consistency.

## Changes to rememberRetained Behavior

The following explanation is based on the code sample below.

```kotlin
@Composable
private fun ConditionalRetainContent(registry: RetainedStateRegistry) {
  CompositionLocalProvider(LocalRetainedStateRegistry provides registry) {
    var showContent by remember { mutableStateOf(false) }
    Column {
      Button(modifier = Modifier.testTag(TAG_BUTTON_HIDE), onClick = { showContent = false }) {
        Text(text = "Hide content")
      }
      Button(modifier = Modifier.testTag(TAG_BUTTON_SHOW), onClick = { showContent = true }) {
        Text(text = "Show content")
      }
      if (showContent) {
        var count by rememberRetained { mutableIntStateOf(0) }
        Button(modifier = Modifier.testTag(TAG_BUTTON_INC), onClick = { count += 1 }) {
          Text(text = "Increment")
        }
        Text(modifier = Modifier.testTag(TAG_RETAINED_1), text = count.toString())
      }
    }
  }
}
```

### Current Behavior

#### Before `saveAll` is called on the registry

* `rememberRetained` added when the condition is `true` will be removed when the condition changes to `false`.
* When the condition changes back to `true` and the same `rememberRetained` is called, the previous value is not retained.

[Test code for this behavior](https://github.com/slackhq/circuit/blob/82748dafb82f941cb0229eb9037ed5e9662b541d/circuit-retained/src/androidInstrumentedTest/kotlin/com/slack/circuit/retained/android/RetainedTest.kt#L400)

#### After `saveAll` is called on the registry

* `rememberRetained` added when the condition is `true` will be removed when the condition changes to `false`.
* Then, `saveAll` is called on the registry.
* When the condition changes back to `true` and the same `rememberRetained` is called, the previous value is retained.

[Test code for this behavior](https://github.com/slackhq/circuit/blob/82748dafb82f941cb0229eb9037ed5e9662b541d/circuit-retained/src/androidInstrumentedTest/kotlin/com/slack/circuit/retained/android/RetainedTest.kt#L424)

### Challenges with the Current Behavior
* The timing of the `saveAll` call on the `RetainedStateRegistry` is unknown from the perspective of lower-level content, making retention depend on whether `saveAll` was called.
* In the case of `remember` / `rememberSaveable`, the node doesn’t retain the previous value when it’s removed and added back, making it easy to expect `rememberRetained` to behave similarly.

Due to these challenges, I propose modifying `rememberRetained` to behave like `remember` / `rememberSaveable`, so that `rememberRetained` does not retain values when it is hidden and reshown in the compose node based on a condition.

## Internal Implementation Changes
Overall, I drew inspiration from the implementation of `SaveableStateHolder` and `rememberSaveable` for these modifications.

The following changes were made to achieve this goal:

### 1. Modify `RetainableSaveableHolder` to always unregister values from the `RetainedStateRegistry` when `onForgotten` is called.
In the previous implementation, if `rememberRetained` was removed from the node and `onForgotten` was called, values were not unregistered from the registry if `canRetain` was `true`. As a result, all values were retained in the registry regardless of whether `rememberRetained` was present in the node when `saveAll` was called.

Now, values are unregistered from the registry when `onForgotten` is called, so only values present in the composition node are retained when `saveAll` is called .

### 2. Change the timing of `saveAll` in `RetainedStateRegistry`

Previously, `saveAll` was called on the `RetainedStateRegistry` within `RetainableSaveableHolder` when `onForgotten` was invoked. However, due to the change in 1, `onForgotten` of all child `rememberRetained` nodes is called before `onForgotten` of `rememberRetained { RetainedStateRegistry() }`, which would result in no values being retained.

Therefore, I referenced the `SaveableStateHolder` implementation to create a separate `DisposableEffect` to run `saveAll`, as shown below:

```kotlin
val parentRegistry = LocalRetainedStateRegistry.current
val registry = rememberRetained(key) { RetainedStateRegistry() }

CompositionLocalProvider(LocalRetainedStateRegistry provides registry) {
    // child content
    ... 
}

DisposableEffect(key, registry) {
    onDispose { 
        registry.saveAll()
        parentRegistry.saveValue(key)
    }
}
```

By structuring it this way, `saveAll` can be called on the `RetainedStateRegistry` before the child content's dispose stage. Thus, even if `onForgotten` is called on the child content's `rememberRetained` and unregisters the value, it will still be retained.

Compared to the existing implementation, this change needs to be applied to all cases where the `RetainedStateRegistry` is redefined in a nested way. To facilitate this process, I made the [`RetainedStateProvider`](https://github.com/vulpeszerda/circuit/blob/50dc04e38158d12f61ac192c14b37d7f84aa5ec1/circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RetainedStateProvider.kt) function public and modified `NavigableCircuitContent` and `PausableState` to use it. (2f2b06fcbb9a10afb6d5905e8c7da6f4e99dd202)



### 3. Change the timing of `saveAll` in `AndroidContinuity`

For the same reason as in 2, `continuityRetainedStateRegistry` also needs to call `saveAll` using `DisposableEffect` after declaring the child content.

#### Example
```kotlin
val registry = continuityRetainedStateRegistry()

CompositionLocalProvider(LocalRetainedStateRegistry provides registry) {
    // child content
    ... 
}

DisposableEffect(registry) {
    onDispose { 
        registry.saveAll()
    }
}
```

However, in Android, `continuityRegistry` should always call `saveAll` when `onStop` is called. Since calling `saveAll` multiple times won’t cause issues, I modified it to call `saveAll` conveniently upon `onStop` or disposal.

```kotlin
@Composable
public fun continuityRetainedStateRegistry(
  key: String = Continuity.KEY,
  factory: ViewModelProvider.Factory = ContinuityViewModel.Factory,
  canRetainChecker: CanRetainChecker = LocalCanRetainChecker.current ?: rememberCanRetainChecker(),
): RetainedStateRegistry {
  @Suppress("ComposeViewModelInjection")
  val vm = viewModel<ContinuityViewModel>(key = key, factory = factory)

  LifecycleStartEffect(vm) {
    onStopOrDispose {
      if (canRetainChecker.canRetain(vm)) {
        vm.saveAll()
      }
    }
  }

  LaunchedEffect(vm) {
    withFrameNanos {}
    // This resumes after the just-composed frame completes drawing. Any unclaimed values at this
    // point can be assumed to be no longer used
    vm.forgetUnclaimedValues()
  }

  return vm
}
```

## Changes to Test Code

### Change 1
In `RetainedTest.kt`, parts where `nestedRegistry` is declared now use `RetainedStateProvider`, as `saveAll` must be manually called.
50dc04e38158d12f61ac192c14b37d7f84aa5ec1

### Change 2
In `NestedRetainWithPushAndPop` and `NestedRetainWithPushAndPopAndCannotRetain`, the tests assume the same value is retained regardless of the `showNestedContent` value, so I set the same key in `RetainedStateProvider` to retain the values. 

However, since these assumptions may change with this PR, it may need verification to ensure correctness.
50dc04e38158d12f61ac192c14b37d7f84aa5ec1

### Change 3
To test `ImpressionEffect`, I modified the function attempting to recreate it. Previously, the condition was first set to `false` to remove the child content and then `saveAll` was performed. In the modified behavior, `saveAll` is performed first and then the child content is removed.

e7cb384988ab98a71b2d65629200694467d4c5c8


## Additional Test Cases for the Reported Issue

I added test cases to cover the issue reported, which show which tests failed with the previous implementation and how they succeed with the modified implementation.

ad1646fd4ef571670705612e6a36f34d0cf1a242, 82748dafb82f941cb0229eb9037ed5e9662b541d


--------------

When I initially reported this issue, I may have been somewhat aggressive due to the unexpected behavior. I apologize if it came across that way. Through working on this modification, I had the chance to explore the internal structure of `rememberRetained` and have come to appreciate how robust and well-designed many parts of the `circuit` library are. I have great respect for the efforts of the main contributors.

After considering various approaches, I believe this change to the behavior of `rememberRetained` is the right direction. However, since the implementation of `rememberRetained` is a core part of `circuit`, there may be differing perspectives on this.

While having this PR approved would be ideal, if there are differing views, I hope we can use this as a foundation for further discussion.

vulpeszerda commented:
Timeout failure occurred when attempting to launch the emulator in the CI. 
 I’m not sure how to proceed from here. 😢

vulpeszerda commented:
@ZacSweers Can you help me about this?

ZacSweers commented:
Please be patient and don't tag maintainers, we will get to it when we get to it!

alexvanyo commented on circuit-retained/src/androidInstrumentedTest/kotlin/com/slack/circuit/retained/android/RetainedTest.kt:
```diff
@@ -696,15 +737,9 @@ private fun NestedRetainWithPushAndPopAndCannotRetain(useKeys: Boolean) {
 
     // Keep the retained state registry around even if showNestedContent becomes false
-    CompositionLocalProvider(LocalCanRetainChecker provides CanRetainChecker.Always) {
+    CompositionLocalProvider(LocalCanRetainChecker provides { false }) {
```
Is this line still necessary?

If `rememberRetained` has its state wiped, from leaving composition separately from either:
- being recreated
- the wrapping `RetainedStateProvider` leaving composition

Then I would also expect a `RetainedStateProvider` that leaves in a similar way to has its state wiped, I think.

vulpeszerda replied:
@alexvanyo 

The state of rememberRetained is wiped in two cases:
1.	When it itself leaves the composition.
2.	When the wrapping RetainedStateProvider leaves the composition.

In this case, the line `CompositionLocalProvider(LocalCanRetainChecker provides { false }) {` is still necessary, as it simulates a special situation where retaining is explicitly denied.

However, I realized a critical issue in my previous implementation based on your comment:
`Then I would also expect a RetainedStateProvider that leaves in a similar way to has its state wiped, I think.`

The problem is that my implementation did not differentiate between the case `when the RetainedStateProvider leaves the composition entirely` vs. `when its key changes`. As a result, the state of the RetainedStateProvider was not cleared when the RetainedStateProvider itself was removed from the composition.

To properly handle this, the RetainedStateProvider needs to distinguish between these two cases (key change vs. removal from composition). I believe this requires a modification similar to how rememberSaveableStateHolder provides a single holder.

I will make the necessary changes and update the code accordingly. Thanks!

vulpeszerda replied:
I fixed above in [these commits](https://github.com/vulpeszerda/circuit/compare/67d156f...9fbc2df)

alexvanyo reviewed (commented):
I agree with the goal of having `rememberRetained` match in behavior with `rememberSaveable` and `remember`: if one of those leaves composition in an `if`, the state is wiped, and I would expect the same for `rememberRetained`.

One thing not clear right now to me still after running some tests: if there is the `if (...) { rememberRetained() }` inside the `RetainedStateProvider`, I think by the same logic, state should also be lost.

However, in my tests, the state is still preserved in that case.

Maybe the `LocalCanRetainChecker` inside `RetainedStateProvider` shouldn't be set to the static `CanRetainChecker.Always`?

vulpeszerda commented on circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RememberRetained.kt:
```diff
@@ -367,13 +340,29 @@ private class RetainableSaveableHolder<T>(
 
   override fun onForgotten() {
-    saveIfRetainable()
+    val v = value
```
Ok I will change this!

vulpeszerda replied:
fixed in 5d30dd71cf17140fe13836b4e8eabc884c133b37

vulpeszerda commented:
@alexvanyo 
Thank you for review. 

However, I couldn't reproduce the test case you mentioned. (I tested with [this](https://github.com/slackhq/circuit/blob/4b20c86d7afcfe0349e6f6e649a7ea0cb0ebba99/circuit-retained/src/androidInstrumentedTest/kotlin/com/slack/circuit/retained/android/RetainedTest.kt#L447))
Could you please share the exact test case? 

> One thing not clear right now to me still after running some tests: if there is the `if (...) { rememberRetained() }` inside the `RetainedStateProvider`, I think by the same logic, state should also be lost.
> 
> However, in my tests, the state is still preserved in that case.

alexvanyo reviewed (commented):
The `RetainedStateHolder` API looks very similar to #1168 as its now mimicking the `SaveableStateHolder` API exactly, so curious to hear from Zac as to how that exploration went previously.

vulpeszerda commented on circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RetainedStateHolder.kt:
```diff
@@ -0,0 +1,79 @@
+      val childRegistry = rememberRetained(key = key) { RetainedStateRegistry() }
+      CompositionLocalProvider(
+        LocalRetainedStateRegistry provides childRegistry,
+        LocalCanRetainChecker provides CanRetainChecker.Always,
```
When creating separate `RetainedStateRegistry` instances within `RetainedStateProvider`, the `CanRetainChecker` propagated to the child has a different meaning from the parent’s `CanRetainChecker`.

**Meaning of the child’s `CanRetainChecker`:**

The child’s `CanRetainChecker` determines whether `rememberRetained` declared under the `RetainedStateProvider` can be retained in cases where the key transitions from A -> B -> A. This behavior should always function correctly even if the parent’s `CanRetainChecker` always returns false, because the `RetainedStateHolder` itself has never been disposed. As long as the `RetainedStateHolder` is not disposed, `rememberRetained` declared under the `RetainedStateProvider` should always be restorable, so it seems appropriate to pass `CanRetainChecker.Always`.

**Meaning of the parent’s `CanRetainChecker`:**

The `CanRetainChecker` provided by the parent of the `RetainedStateHolder` determines whether the internal data of the `RetainedStateHolder` should be retained when the screen is recreated.

-----------

For these reasons, I think the existing code also passed `CanRetainChecker.Always` when creating a new registry. Following the intention of the original code, I believe redefining it as `CanRetainChecker.Always` in my code is necessary.

> I tried removing it, and all of the tests still pass, so maybe this can now be removed?

I think the current test cases might be insufficient. I expect a difference in behavior for the following scenario:

1. The parent’s CanRetainChecker always returns false.
2. A RetainedStateHolder is defined.
3. The key in holder.RetainedStateProvider transitions from A -> B -> A.

When redefined with CanRetainChecker.Always -> rememberRetained inside holder.RetainedStateProvider is preserved.

Without redefined ->
rememberRetained inside holder.RetainedStateProvider is not preserved.

alexvanyo replied:
Thank you for the explanation - I'm understanding now the rationale here for `Always`.

This change swaps who is responsible for performing the retaining:

It is now the parent's responsibility to save children `rememberRetained`, moving away from having it be the `rememberRetained`'s responsibility to tell its parent to save it.

As a result of this, specifying `LocalCanRetainChecker provides CanRetainChecker.Always` for the children is the parent signifying that as a custom retaining scope, when the parent chooses to save, at that moment all children present should be saved.

vulpeszerda commented on circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RetainedStateHolder.kt:
```diff
@@ -0,0 +1,79 @@
+ * values. Each [RetainedStateProvider] maintains a unique retainedStateRegistry for each key,
+ * allowing it to save and restore states.
+ */
+public interface RetainedStateHolder {
```
Sure do! fb672cfc365638fa5450adbf427f9b2794c46e86

ZacSweers commented:
> The RetainedStateHolder API looks very similar to https://github.com/slackhq/circuit/pull/1168 as its now mimicking the SaveableStateHolder API exactly, so curious to hear from Zac as to how that exploration went previously.

Still want it! It essentially stalled because I hadn't picked it back up again. If that's something that naturally can fall out of this work then that'd be dope 👌

vulpeszerda commented:
Is there any additional work I need to do to get my changes reviewed?

ZacSweers commented:
Sorry I think we were waiting for a resolution on my last comment?

> Still want it! It essentially stalled because I hadn't picked it back up again. If that's something that naturally can fall out of this work then that'd be dope 👌

Is that something that falls out of this PR? Should it?

vulpeszerda commented:
Ah, I thought contributors were busy with reviews since it’s the end of the year.

I’ll analyze issue #1168 in more detail and review whether replacing it with the `RetainedStateHolder` I added in my work would resolve it. Also, since we’re on the topic, it would be great to include the test cases added in #1168 here as well. 

I’ll work on that.

ZacSweers commented:
Sounds good!

alexvanyo commented on circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RetainedStateHolder.kt:
```diff
@@ -0,0 +1,86 @@
+  override fun RetainedStateProvider(key: String, content: @Composable (() -> Unit)) {
+    CompositionLocalProvider(LocalRetainedStateRegistry provides this) {
+      val canRetainChecker = LocalCanRetainChecker.current ?: CanRetainChecker.Always
+      val childRegistry = rememberRetained(key = key) { RetainedStateRegistry() }
```
This is minor, but I think specifying the `key` here as the `key` to `rememberRetained` has the downside of polluting the key space with a risk of conflicts: if you had a `RetainedStateProvider` in the `RetainedStateRegistry` scope with overlapping key sets, then this `rememberRetained` could conflict.

I think a safer way to do this is to add the composite key hash, either explicitly, or do what `SaveableStateHolder` does and doi it implicitly by adding a `ReusableContent(key)` around this whole content.

vulpeszerda replied:
@alexvanyo yes right. 
I confused that rememberRetained would add composite key has as prefix even though the key provided. 
Adding composite key hash as prefix would be easy fix. But adding `ReusableContent(key)` would be more proper fix for considering other content (non circuit related) reuse. 
I haven’t had much free time, so I haven’t been able to work on what Zac mentioned earlier yet. I’ll try to incorporate this as well tomorrow and next week.

vulpeszerda replied:
@alexvanyo,

After much consideration and testing, I’ve decided to keep the current implementation that passes `rememberRetained(key = key)` here. There are several reasons for this:

#### 1. RetainedStateHolder usage
`RetainedStateHolder` constructs its own `RetainedStateRegistry` and uses it only once to store a child registry, which makes conflicts highly unlikely. The only potential conflict scenario would be if holder.RetainedStateProvider is used in multiple places, as shown below.

```kotlin
val holder = rememberRetainedStateHolder()
holder.RetainedStateProvider("key1") { ... }
holder.RetainedStateProvider("key1") { ... }
```

#### 2. Handling removeState
When `removeState` is called and the `holder.RetainedStateProvider` for the corresponding key is already disposed (meaning that child registry remains retained in the parent registry), the parent registry needs to remove that retained child registry. However, if we rely on a composite key hash instead of the direct key, we cannot determine which composite key hash corresponds to the original key.


```kotlin
override fun removeState(key: String) {
  val canRetainChecker = canRetainCheckers[key]
  if (canRetainChecker != null) {
    canRetainChecker.shouldSave = false
  } else {
    registry.consumeValue(key) // <--- this case
  }
}
```

#### 3. SaveableStateHolder consistency
Internally, `SaveableStateHolder` also uses the key argument to distinguish between different data sets.

--------------

For these reasons, I will continue passing `rememberRetained(key = key)` as is. 
However, since I appreciate the usability of `ReusableContent` in `SaveableStateHolder`, I’ve updated that part accordingly in 16cf579170bd954ff272122c4fee25e1f43e030c.

alexvanyo replied:
Got it, that makes sense to me for usage. Thank you for writing up the details here and evaluating each option thoroughly!

vulpeszerda commented:
I’ve cherry-picked the `RetainedStateHolderTest` from #1168 and integrated it into this branch. The tests pass with the `RetainedStateHolder` implementation in this PR, and everything appears to be working fine.

To use the existing circuit’s `RetainedStateRegistry` without significantly modifying its implementation, I had to implement `RetainedStateHolder` differently than the one in previous PR.

From my understanding, the issue addressed in #1168 is resolved by using this `RetainedStateHolder`, so I went ahead and applied those changes in commit 9714ce3b4ee900179c570f6ba94dfbe4ba01c79c.

It looks like all the issues have been resolved. If there’s anything else you’d like me to work on, please let me know!

ZacSweers commented on circuit-retained/src/androidInstrumentedTest/kotlin/com/slack/circuit/retained/android/RetainedStateRestorationTester.kt:
```diff
@@ -0,0 +1,138 @@
+ */
+// TODO recreate for more realism? Need to save the content function to do that, or call it after
+// TODO make this available in test utils?
+class RetainedStateRestorationTester(private val composeTestRule: ComposeContentTestRule) {
```
nit: not for this PR, but do you think we should make this available as a test utility somewhere in circuit-test?

vulpeszerda replied:
Moving it to circuit-test could make it more accessible for others, similar to `StateRestorationTester`, which seems like a good idea. But I’m not quite sure how frequently `emulateRetainedInstanceStateRestore` would be used.

ZacSweers replied:
yeah we can save that for another time!

ZacSweers commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/RetainedStateHolder.kt:
```diff
@@ -0,0 +1,106 @@
+import com.slack.circuit.retained.RetainedValueProvider
+import com.slack.circuit.retained.rememberRetained
+
+/** Copy of [RetainedStateHolder] to return content value */
```
nit: do you think we can/should reconcile these into one API?

vulpeszerda replied:
I think it would be good to consolidate them, but there are a few considerations:

1. Just like `SaveableStateHolder` in #1451 , we need to investigate further whether converting a `RetainedStateHolder` without a return value into one with a return value would cause any issues.

2. The `RetainedStateHolder` without a return value internally uses `ReusableContent(key)`, while the one with a return value uses `key(key)`. I’m not entirely sure if using `ReusableContent(key)` in the case of a return value would cause any issues.

ZacSweers reviewed (approved):
Thanks a bunch! @alexvanyo is going to take one more look today as well then I think we can move ahead with this. Thanks a bunch for your patience in this. Josh is on vacation and we've had a bit of a chaotic month to start the new year

alexvanyo reviewed (approved):
Thank you for the work here and investigating the different options! The resulting API and usage matches my expectations more, and is more consistent with the `SaveableStateProvider` setup.

vulpeszerda commented on samples/star/src/commonMain/kotlin/com/slack/circuit/star/home/HomeScreen.kt:
```diff
@@ -127,12 +129,19 @@ fun HomeContent(state: HomeScreen.State, modifier: Modifier = Modifier) =
+          )
+        }
+      }
+      contentComposed = true
```
Nice. I might have made a mistake while resolving the conflict during the merge. 😢  
Fixed!

vulpeszerda commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/NavigableCircuitContent.kt:
```diff
@@ -190,21 +191,16 @@ private fun <R : Record> buildCircuitContentProviders(
         // Now provide a new registry to the content for it to store any retained state in,
         // along with a retain checker which is always true (as upstream registries will
         // maintain the lifetime), and the other provided values
```
Um.. yes I changed a little bit.

ZacSweers commented:
Sorry one last request - would you mind doing a writeup of this change into the CHANGELOG.md file? It can be a longer/more detailed section to explain it since it's more significant of a change :). Let's make sure we highlight both the behavior changes and the new APIs.

alexvanyo commented on circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RetainedStateHolder.kt:
```diff
@@ -0,0 +1,109 @@
+          LocalCanRetainChecker provides CanRetainChecker.Always,
+          content = content,
+        )
+        DisposableEffect(Unit) {
```
Something very subtle I ran into experimenting with this [change](https://github.com/alexvanyo/composelife/pull/2305): I think this needs to have both `entryCanRetainChecker` and `childRegistry` as keys, or alterantively using `rememberUpdatedState` with them to avoid operating on stale references to old `childRegistry`s or `entryCanRetainChecker`s.

This is exacerbated by #1919 - if the `CanRetainChecker` instance changes, then this can get into a bad state where the child registry gets a new instance, and this disposable effect operates on the stale old instance.

vulpeszerda replied:
Sorry for bringing up a difficult discussion again at the end of a long task, but to properly address the issue Alex raised, I think I need to understand how the Main Contributors think about the following concerns.

#### Should `CanRetainChecker` only affect `RetainedStateRegistry`?
`CanRetainChecker` originally seems to have existed for `RetainedStateRegistry`. However, since it is provided via CompositionLocal, the following code can occur:

```kotlin
val registry = remember { RetainedStateRegistry() }

CompositionLocalProvider(LocalRetainedStateRegistry provides registry) {

    CompositionLocalProvider(LocalCanRetainChecker provides remember { CanRetainChecker { false }}) {
          val value = rememberRetained { ... }
    }
}
```

In this case, when `registry.saveAll()` is called, the value should / should not be saved?

I believe that `LocalCanRetainChecker` should only affect to RetainedStateRegistry level, not for individual rememberRetained. (which means the value should be saved)

ZacSweers replied:
I think right now we need to check in rememberRetained still. We could revisit being able to check it solely in RetainedStateRegistry but I think that's outside the scope of this PR maybe?

vulpeszerda replied:
Yes right. I agree. For this update, I’ll make sure `LocalCanRetainChecker` continues to work as expected and still affects `rememberRetained`.

However, the changes I made earlier caused an issue where `LocalCanRetainChecker` no longer influenced `rememberRetained`, so I’ve fixed that.

ZacSweers replied:
Sounds good. Don't forget the changelog writeup!

vulpeszerda replied:
Of course! 🙂 
I was actually writing a more detailed explanation of the changes I just uploaded.
But it’s taking a bit of time, so it’d be great if you could review it tomorrow or the day after!

vulpeszerda replied:
Alex, since the code in `RetainedStateHolder` that referenced parent `CanRetainChecker` has been removed, we no longer need to handle the scenario of CanRetainChecker changes separately.

To elaborate, in your #1920 you addressed changes in `CanRetainChecker` within `rememberRetained`. In my revised approach, whether the child registry is saved in `RetainedStateHolder` is automatically determined by the `CanRetainChecker` applied to `rememberRetained` in below
```kotlin
rememberRetained(key = key) { RetainedStateRegistry() }
```

alexvanyo replied:
I think I see - so the latest changes now fully rely upon `CanRetainChecker` being used in `rememberRetained`, instead of it also being checked inside `RetainedStateHolder`. Makes sense to me!

vulpeszerda commented:
As I mentioned above, 
I’ve added a test case for the scenario Alex originally commented on, where CanRetainChecker changes. During this process, I discovered and fixed unintended behavior change where `LocalCanRetainChecker` was not affecting `rememberRetained`.

Additionally, as part of this fix, I simplified the model by removing the unnecessary `EntryCanRetainChecker`. In the previous implementation of `RetainedStateHolder`, the parent scope’s `LocalCanRetainChecker` was already influencing the child registry’s `canBeRetained`, so now it only references the value for `shouldSave`.

As the final step of this long task, I’ll update the PR description to reflect the latest changes and add the changelog as requested by Zac

ZacSweers commented on CHANGELOG.md:
```diff
@@ -3,6 +3,40 @@ Changelog
 
 **Unreleased**
 --------------
+### Behaviour Change: rememberRetained
```
Great writeup!
